// UserManagement/index.js
function arrayRemove(arr, value) {

    return arr.filter(function(ele){
        return ele != value;
    });
 
}

function printFile(){
    let i = 0;
    for (let element of file ){ 
        console.log(i+" : "+element.id);
        i ++;
    }
}

// Variables
var clients = [];// Tous les clients
var file = [];// Clients en attente d'adversaires
var games = new Map();// Clients en games


// Fonction recherche de fonction 
searchplayer = function searchplayer(socket){
    file.push(socket);
    console.log("Dans la file attent :");
    printFile();
    let find = false;
    for (let element of file ){ 
        if(element.id != socket.id){
            games.set(element,socket);
            games.set(socket,element);
            find = true;
            console.log("joueur trouvée !");
            console.log("connection entre :"+games.get(socket).id+" et "+socket.id);
            break;
        }
    };
    return find ? true : false;
}

updateWaiting = function updateWaiting(socket){
    file = arrayRemove(file,socket);
    file = arrayRemove(file,games.get(socket));
    console.log("FileModule :"+file)
    for (let element of file ){ 
        console.log(element.id);
    }
    return file;
}

disconnectPlayers = function disconnectPlayers(socket,games){
    games.delete(socket);
    games.delete(games.get(socket));
}

userconnection = function userconnection(socket){
    clients.push(socket)
    console.log('Utilisateur connecté Id :'+socket.id);
}

addToWaitingFile = function addToWaitingFile(socket){
    console.log(socket.id+ " à la recherche d'un joueur ...");
    clients.push(socket);
}

sendMessageToUser = function sendMessageToUser(socket,message){
    socket.emit('receiveMessage',message);
}
sendMessageToGame = function sendMessageToUser(socket,message){
    games.get(socket).emit('receiveMessage',message+socket.id)
    socket.emit('receiveMessage',message+socket.id);
}
// exports the variables and functions above so that other modules can use them
module.exports = userconnection;
module.exports = sendMessageToUser;
module.exports = sendMessageToGame;
module.exports = addToWaitingFile;
module.exports = searchplayer;
module.exports = updateWaiting;
