// UserManagement/index.js
// Variables
var clients = [];// Tous les clients connectés aux serveurs actuellement
var file = [];// Clients en attente d'adversaires
var games = new Map();// Clients en games


// Fonction recherche de fonction 
searchplayer = function searchplayer(socket){
    file.push(socket);
    console.log("La file d'attente :");
    printFile();
    let find = false;
    for ( let element of file ){ 
        if( element.id != socket.id ){
            games.set(element,socket);
            games.set(socket,element);
            find = true;
            console.log( "joueur trouvée !" );
            console.log( "connection entre :"+ games.get(socket).id +" et "+ socket.id );
            break;
        }
    };
    return find ? true : false;
}

// True : joueur en game
isOnGame = function isOnGame(socket){
    if ( games.get(socket) != undefined ){ return true }else{ return false;}
}

// Mise à jour de la file d'attente
updateWaitingFile = function updateWaitingFile(socket){
    file = arrayRemove(file,socket);
    file = arrayRemove(file,games.get(socket));
    console.log("FileModule :"+file)
    for ( let element of file ){ 
        console.log(element.id);
    }
    return file;
}

// Fin de partie ...Changement de nomfunction?
disconnectPlayers = function disconnectPlayers(socket,games){
    games.delete(socket);
    games.delete(games.get(socket));
    // TODO : gestion de fin de partie
}

// Connection d'un joueur
userconnection = function userconnection(socket){
    clients.push(socket);// Ajout de l'utilsateurs aux tableaux des clients connectés
    console.log('Utilisateur connecté Id :'+socket.id);
    // TODO : Gestion de reconnection
}

// Ajout à la file d'attente
addToWaitingFile = function addToWaitingFile(socket){
    console.log( socket.id + " à la recherche d'un joueur ..." );
    clients.push(socket);
}

// Envoie un message à un autre utilisateur
sendMessageToUser = function sendMessageToUser(socket,message){
    socket.emit('receiveMessage',message);
}

// Envoie un message aux joueurs de la partie
sendMessageToGame = function sendMessageToGame(socket,message){
    games.get(socket).emit('receiveMessage',message+socket.id)
    socket.emit('receiveMessage',message+games.get(socket).id);
}

// Un utilisateur se rend
surrender = function surrender(socket){
    console.log(socket.id+" : se rend. "+"On emit la victoire à :"+games.get(socket).id)
    // On notifie les deux joueurs true pour victoire
    games.get(socket).emit('receiveEndGame',true);
    socket.emit('receiveEndGame',false);
    // On enlève les utilisateurs du tableau games
    games = arrayRemove(games,socket);
    games = arrayRemove(games,games.get(socket));
}

// Modification de pseudo
setPseudo = function setPseudo(pseudo,socket){
    console.log(socket.id+" change son pseudo en "+pseudo);
    // Si l'utilisateur est en game, on notifie son adversaire
    if ( isOnGame(socket) ) games.get(socket).emit('receivePseudo',"Votre adversaire à changé son pseudo en :"+pseudo);
    socket.emit('receivePseudo',"Vous changez votre pseudo en :"+pseudo);
    // TODO : save pseudo local storage
}

// exports the variables and functions above so that other modules can use them
module.exports = userconnection;
module.exports = sendMessageToUser;
module.exports = sendMessageToGame;
module.exports = addToWaitingFile;
module.exports = searchplayer;
module.exports = updateWaitingFile;
module.exports = surrender;
module.exports = setPseudo;


// suprime un élément d'un tableau
function arrayRemove(arr, value) {

    return arr.filter(function(ele){
        return ele != value;
    });
 
}
// Affiche un tableau 
function printFile(){
    let i = 0;
    for (let element of file ){ 
        console.log(i+" : "+element.id);
        i ++;
    }
}
